    In this notebook, we apply text mining / NLP techniques to extract insights from textual features.Then we will try to     predict restaurant ratings using ML techniques..

In [300]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### so that u dont have warnings
from warnings import filterwarnings
filterwarnings('ignore')

In [301]:
# Reading restaurants data
data_path = 'F:\Data Science projects\EDA_Projects\zomato-bangalore-restaurants\zomato.csv'
from pandas import read_csv

In [302]:
df= read_csv(data_path)

# Results
print(f'Dataset shape: {df.shape}')
df.head()

Dataset shape: (51717, 17)


url  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes                             phone  \
0          Yes        Yes  4.1/5    775    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5    787                      080 41714161   
2          Yes         No  3.8/5    918                    +91 9663487993   
3           No         No  3.7/5     88                    +91 9620009302   
4           No         No  3.8/5    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        reviews_list menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

  listed_in(type) listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

    An overview from the data

In [303]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 17 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   url                          51717 non-null  object
 1   address                      51717 non-null  object
 2   name                         51717 non-null  object
 3   online_order                 51717 non-null  object
 4   book_table                   51717 non-null  object
 5   rate                         43942 non-null  object
 6   votes                        51717 non-null  int64 
 7   phone                        50509 non-null  object
 8   location                     51696 non-null  object
 9   rest_type                    51490 non-null  object
 10  dish_liked                   23639 non-null  object
 11  cuisines                     51672 non-null  object
 12  approx_cost(for two people)  51371 non-null  object
 13  reviews_list                 51

In [304]:
df.isnull().sum()

url                                0
address                            0
name                               0
online_order                       0
book_table                         0
rate                            7775
votes                              0
phone                           1208
location                          21
rest_type                        227
dish_liked                     28078
cuisines                          45
approx_cost(for two people)      346
reviews_list                       0
menu_item                          0
listed_in(type)                    0
listed_in(city)                    0
dtype: int64

In [305]:
feature_na=[feature for feature in df.columns if df[feature].isnull().sum()>0]
feature_na

['rate',
 'phone',
 'location',
 'rest_type',
 'dish_liked',
 'cuisines',
 'approx_cost(for two people)']

In [306]:
#% of missing values
import numpy as np
for feature in feature_na:
    print('{} has {} % missing values'.format(feature,np.round(df[feature].isnull().sum()/len(df)*100,4)))

rate has 15.0337 % missing values
phone has 2.3358 % missing values
location has 0.0406 % missing values
rest_type has 0.4389 % missing values
dish_liked has 54.2916 % missing values
cuisines has 0.087 % missing values
approx_cost(for two people) has 0.669 % missing values


In [307]:
# Making a copy of the data to work on
data = df.copy()

In [308]:
# Dropping duplicates -

In [309]:
data.groupby(["name", "address"]).agg({"listed_in(type)" : list})

listed_in(type)
name                                         address                                                                                              
#FeelTheROLL                                 Opposite Mantri Commercio, Outer Ring Road, Dev...                               [Delivery, Delivery]
#L-81 Cafe                                   Sector 6, HSR Layout, HSR                           [Delivery, Dine-out, Delivery, Dine-out, Deliv...
#Vibes Restro                                Marasur Gate, Chandapura - Anekal Road, Near Al...                       [Buffet, Delivery, Dine-out]
#refuel                                      7, Ground Floor, RR Commercial Complex, Akshay ...                        [Cafes, Delivery, Dine-out]
'Brahmins' Thatte Idli                       19, 1st main, 2nd cross, 3rd stage, 3rd block, ...                                         [Dine-out]
...                                                                                                                                            ...
nu.tree                                      47/7 First floor, DoddaThogur Road, Near Velank...                                         [Delivery]
                                             Ground Floor-Lobby Area, Brigade IRV Centre, Na...                     [Delivery, Delivery, Delivery]
re:cess - Hilton Bangalore Embassy GolfLinks Hilton Bangalore Embassy GolfLinks, Embassy Gol...                [Dine-out, Dine-out, Pubs and bars]
repEAT Hub                                   67/4, Bhoganahalli Road, JCR Layout, Panathur, ...                               [Delivery, Dine-out]
sCoolMeal                                    197/293-3, 32nd Main, 1st Stage, BTM, Bangalore     [Delivery, Delivery, Delivery, Delivery, Deliv...

[12499 rows x 1 columns]

In [310]:
## it means we have to aggregate data of listed_in(type) into list
grouped=data.groupby(["name", "address"]).agg({"listed_in(type)" : list})

In [311]:
df.columns

Index(['url', 'address', 'name', 'online_order', 'book_table', 'rate', 'votes',
       'phone', 'location', 'rest_type', 'dish_liked', 'cuisines',
       'approx_cost(for two people)', 'reviews_list', 'menu_item',
       'listed_in(type)', 'listed_in(city)'],
      dtype='object')

In [312]:
## merge your both the dataframe
newdata = pd.merge(grouped, data, on = (["name", "address"]))

In [313]:
newdata.columns

Index(['name', 'address', 'listed_in(type)_x', 'url', 'online_order',
       'book_table', 'rate', 'votes', 'phone', 'location', 'rest_type',
       'dish_liked', 'cuisines', 'approx_cost(for two people)', 'reviews_list',
       'menu_item', 'listed_in(type)_y', 'listed_in(city)'],
      dtype='object')

In [314]:
newdata['listed_in(type)_y'].dtype

dtype('O')

In [315]:
newdata.shape

(51717, 18)

In [316]:
newdata.drop_duplicates(subset = ["name", "address", "listed_in(type)_y"], inplace = True)

In [317]:
newdata.shape

(20915, 18)

In [318]:
newdata.head()

name                                            address  \
0    #FeelTheROLL  Opposite Mantri Commercio, Outer Ring Road, De...   
2      #L-81 Cafe                          Sector 6, HSR Layout, HSR   
3      #L-81 Cafe                          Sector 6, HSR Layout, HSR   
11  #Vibes Restro  Marasur Gate, Chandapura - Anekal Road, Near A...   
12  #Vibes Restro  Marasur Gate, Chandapura - Anekal Road, Near A...   

                                    listed_in(type)_x  \
0                                [Delivery, Delivery]   
2   [Delivery, Dine-out, Delivery, Dine-out, Deliv...   
3   [Delivery, Dine-out, Delivery, Dine-out, Deliv...   
11                       [Buffet, Delivery, Dine-out]   
12                       [Buffet, Delivery, Dine-out]   

                                                  url online_order book_table  \
0   https://www.zomato.com/bangalore/feeltheroll-b...           No         No   
2   https://www.zomato.com/bangalore/l-81-cafe-hsr...          Yes         No   
3   https://www.zomato.com/bangalore/l-81-cafe-hsr...          Yes         No   
11  https://www.zomato.com/bangalore/vibes-restro-...           No         No   
12  https://www.zomato.com/bangalore/vibes-restro-...           No         No   

     rate  votes                             phone         location  \
0   3.4/5      7  +91 9108342079\r\n+91 9886117901        Bellandur   
2   3.9/5     48                    +91 9986210891              HSR   
3   3.9/5     48                    +91 9986210891              HSR   
11    NaN      0                    +91 9538911285  Electronic City   
12    NaN      0                    +91 9538911285  Electronic City   

        rest_type dish_liked                       cuisines  \
0     Quick Bites        NaN                      Fast Food   
2     Quick Bites    Burgers           Fast Food, Beverages   
3     Quick Bites    Burgers           Fast Food, Beverages   
11  Casual Dining        NaN  Continental, Chinese, Italian   
12  Casual Dining        NaN  Continental, Chinese, Italian   

   approx_cost(for two people)  \
0                          200   
2                          400   
3                          400   
11                         700   
12                         700   

                                         reviews_list menu_item  \
0   [('Rated 5.0', "RATED\n  Had an egg chicken ro...        []   
2   [('Rated 4.0', 'RATED\n  This little cafe is s...        []   
3   [('Rated 4.0', 'RATED\n  This little cafe is s...        []   
11  [('Rated 5.0', "RATED\n  Great service and don...        []   
12  [('Rated 5.0', "RATED\n  Great service and don...        []   

   listed_in(type)_y  listed_in(city)  
0           Delivery        Bellandur  
2           Delivery              BTM  
3           Dine-out              BTM  
11            Buffet  Electronic City  
12          Delivery  Electronic City

In [319]:
## resetting your index
newdata = newdata.reset_index(drop = True)

In [320]:
newdata.head()

name                                            address  \
0   #FeelTheROLL  Opposite Mantri Commercio, Outer Ring Road, De...   
1     #L-81 Cafe                          Sector 6, HSR Layout, HSR   
2     #L-81 Cafe                          Sector 6, HSR Layout, HSR   
3  #Vibes Restro  Marasur Gate, Chandapura - Anekal Road, Near A...   
4  #Vibes Restro  Marasur Gate, Chandapura - Anekal Road, Near A...   

                                   listed_in(type)_x  \
0                               [Delivery, Delivery]   
1  [Delivery, Dine-out, Delivery, Dine-out, Deliv...   
2  [Delivery, Dine-out, Delivery, Dine-out, Deliv...   
3                       [Buffet, Delivery, Dine-out]   
4                       [Buffet, Delivery, Dine-out]   

                                                 url online_order book_table  \
0  https://www.zomato.com/bangalore/feeltheroll-b...           No         No   
1  https://www.zomato.com/bangalore/l-81-cafe-hsr...          Yes         No   
2  https://www.zomato.com/bangalore/l-81-cafe-hsr...          Yes         No   
3  https://www.zomato.com/bangalore/vibes-restro-...           No         No   
4  https://www.zomato.com/bangalore/vibes-restro-...           No         No   

    rate  votes                             phone         location  \
0  3.4/5      7  +91 9108342079\r\n+91 9886117901        Bellandur   
1  3.9/5     48                    +91 9986210891              HSR   
2  3.9/5     48                    +91 9986210891              HSR   
3    NaN      0                    +91 9538911285  Electronic City   
4    NaN      0                    +91 9538911285  Electronic City   

       rest_type dish_liked                       cuisines  \
0    Quick Bites        NaN                      Fast Food   
1    Quick Bites    Burgers           Fast Food, Beverages   
2    Quick Bites    Burgers           Fast Food, Beverages   
3  Casual Dining        NaN  Continental, Chinese, Italian   
4  Casual Dining        NaN  Continental, Chinese, Italian   

  approx_cost(for two people)  \
0                         200   
1                         400   
2                         400   
3                         700   
4                         700   

                                        reviews_list menu_item  \
0  [('Rated 5.0', "RATED\n  Had an egg chicken ro...        []   
1  [('Rated 4.0', 'RATED\n  This little cafe is s...        []   
2  [('Rated 4.0', 'RATED\n  This little cafe is s...        []   
3  [('Rated 5.0', "RATED\n  Great service and don...        []   
4  [('Rated 5.0', "RATED\n  Great service and don...        []   

  listed_in(type)_y  listed_in(city)  
0          Delivery        Bellandur  
1          Delivery              BTM  
2          Dine-out              BTM  
3            Buffet  Electronic City  
4          Delivery  Electronic City

In [321]:
newdata.shape

(20915, 18)

In [322]:
df['rate'].unique() 

array(['4.1/5', '3.8/5', '3.7/5', '3.6/5', '4.6/5', '4.0/5', '4.2/5',
       '3.9/5', '3.1/5', '3.0/5', '3.2/5', '3.3/5', '2.8/5', '4.4/5',
       '4.3/5', 'NEW', '2.9/5', '3.5/5', nan, '2.6/5', '3.8 /5', '3.4/5',
       '4.5/5', '2.5/5', '2.7/5', '4.7/5', '2.4/5', '2.2/5', '2.3/5',
       '3.4 /5', '-', '3.6 /5', '4.8/5', '3.9 /5', '4.2 /5', '4.0 /5',
       '4.1 /5', '3.7 /5', '3.1 /5', '2.9 /5', '3.3 /5', '2.8 /5',
       '3.5 /5', '2.7 /5', '2.5 /5', '3.2 /5', '2.6 /5', '4.5 /5',
       '4.3 /5', '4.4 /5', '4.9/5', '2.1/5', '2.0/5', '1.8/5', '4.6 /5',
       '4.9 /5', '3.0 /5', '4.8 /5', '2.3 /5', '4.7 /5', '2.4 /5',
       '2.1 /5', '2.2 /5', '2.0 /5', '1.8 /5'], dtype=object)

In [323]:
df['rate'].dtype

dtype('O')

In [324]:
# Transforming the ratings column 

newdata["rating"] = newdata["rate"].str[:3] # Extracting the first three characters of each string in "rate"

In [325]:
newdata["rating"].replace('NEW',0,inplace=True)
newdata["rating"].replace('-',0,inplace=True)

In [326]:
# Converting ratings to a numeric column so we can discretize it
newdata["rating"] = pd.to_numeric(newdata["rating"])

In [327]:
# Discretizing the ratings into a categorical feature with 4 levels
## ie if less than 3, assign 0
## ie if less than 3.5, assign 1
## ie if less than 4.0, assign 2
## ie if less than 5.o assign 3

## We can think of these as 0-Very Low, 1-Low, 2-Medium and 3-High.

In [328]:
newdata["rating"].head()

0    3.4
1    3.9
2    3.9
3    NaN
4    NaN
Name: rating, dtype: float64

In [329]:
newdata['rating'].unique()

array([3.4, 3.9, nan, 3.7, 3.2, 3.5, 4.6, 4.1, 4.2, 4.3, 4. , 3.1, 3.6,
       3.3, 3. , 4.4, 4.5, 0. , 3.8, 2.7, 2.8, 2.9, 2.5, 4.9, 4.7, 4.8,
       1.8, 2.4, 2.1, 2.2, 2.6, 2.3, 2. ])

In [330]:
newdata.isnull().sum()

name                               0
address                            0
listed_in(type)_x                  0
url                                0
online_order                       0
book_table                         0
rate                            3383
votes                              0
phone                            399
location                          12
rest_type                        106
dish_liked                     11999
cuisines                          23
approx_cost(for two people)       98
reviews_list                       0
menu_item                          0
listed_in(type)_y                  0
listed_in(city)                    0
rating                          3383
dtype: int64

In [331]:
newdata.shape

(20915, 19)

In [357]:
## bins = [1, 3.25, 4.1, 5.0]
## it means if it is b/w 1 to 3.25, assign label as 0
## if b/w 3.25 to 4.1, assign label as 1 & so on------
newdata["rating"]=pd.cut(newdata["rating"], bins = [1, 3.25, 4.1, 5.0], labels = ["0", "1", "2"])

In [ ]:
# Visualizing the rating class distribution
plt.figure(figsize = (10, 5))
sns.countplot(newdata["rating"])


In [229]:
# Summary statistics
newdata.describe(include = "all")

name        address     listed_in(type)_x  \
count             20915          20915                 20915   
unique             8792          11495                   927   
top     Cafe Coffee Day  Delivery Only  [Delivery, Dine-out]   
freq                 78             36                  2418   
mean                NaN            NaN                   NaN   
std                 NaN            NaN                   NaN   
min                 NaN            NaN                   NaN   
25%                 NaN            NaN                   NaN   
50%                 NaN            NaN                   NaN   
75%                 NaN            NaN                   NaN   
max                 NaN            NaN                   NaN   

                                                      url online_order  \
count                                               20915        20915   
unique                                              20915            2   
top     https://www.zomato.com/bangalore/dessert-bar-w...          Yes   
freq                                                    1        11605   
mean                                                  NaN          NaN   
std                                                   NaN          NaN   
min                                                   NaN          NaN   
25%                                                   NaN          NaN   
50%                                                   NaN          NaN   
75%                                                   NaN          NaN   
max                                                   NaN          NaN   

       book_table   rate         votes         phone    location    rest_type  \
count       20915  17532  20915.000000         20516       20903        20809   
unique          2     61           NaN         11999          93           93   
top            No  3.9/5           NaN  080 43334333  Whitefield  Quick Bites   
freq        18615   1181           NaN            93        1462         8007   
mean          NaN    NaN    246.105427           NaN         NaN          NaN   
std           NaN    NaN    741.954538           NaN         NaN          NaN   
min           NaN    NaN      0.000000           NaN         NaN          NaN   
25%           NaN    NaN      5.000000           NaN         NaN          NaN   
50%           NaN    NaN     34.000000           NaN         NaN          NaN   
75%           NaN    NaN    167.000000           NaN         NaN          NaN   
max           NaN    NaN  16345.000000           NaN         NaN          NaN   

       dish_liked      cuisines approx_cost(for two people) reviews_list  \
count        8916         20892                       20817        20915   
unique       4502          2630                          70        12660   
top       Biryani  North Indian                         300           []   
freq           78          1174                        3005         3524   
mean          NaN           NaN                         NaN          NaN   
std           NaN           NaN                         NaN          NaN   
min           NaN           NaN                         NaN          NaN   
25%           NaN           NaN                         NaN          NaN   
50%           NaN           NaN                         NaN          NaN   
75%           NaN           NaN                         NaN          NaN   
max           NaN           NaN                         NaN          NaN   

       menu_item listed_in(type)_y listed_in(city) rating  
count      20915             20915           20915  16391  
unique      3693                 7              30      3  
top           []          Delivery             BTM      1  
freq       16766              9355            2379  13130  
mean         NaN               NaN             NaN    NaN  
std          NaN               NaN             NaN    NaN  
min          NaN               NaN             NaN    N

In [33]:
newdata['reviews_list'][0]

'[(\'Rated 5.0\', "RATED\\n  Had an egg chicken roll and a paneer roll... Really yummy... A must visit place... It would be good if you guys can keep the shop open on Sunday\'s as well.. Good luck... :-)"), (\'Rated 5.0\', \'RATED\\n  Not just the Roll but the filling tastes great. I would highly recommend to others to try their rolls. Owners hospitality is also good and very friendly.\'), (\'Rated 4.5\', \'RATED\\n  Very nice place complete value for money ? Highly recommend.Must visit for any foodie . I would recommend the egg chicken roll . #feeltheroll\'), (\'Rated 5.0\', \'RATED\\n  Had an amazing mouth-watering ?chicken roll?Worth every bite..A must try for every foodie?A variety of rolls and sandwiches are available as well .. A budding rafflesia ?\')]'

In [34]:
rev=newdata['reviews_list'][0].lower()
rev

'[(\'rated 5.0\', "rated\\n  had an egg chicken roll and a paneer roll... really yummy... a must visit place... it would be good if you guys can keep the shop open on sunday\'s as well.. good luck... :-)"), (\'rated 5.0\', \'rated\\n  not just the roll but the filling tastes great. i would highly recommend to others to try their rolls. owners hospitality is also good and very friendly.\'), (\'rated 4.5\', \'rated\\n  very nice place complete value for money ? highly recommend.must visit for any foodie . i would recommend the egg chicken roll . #feeltheroll\'), (\'rated 5.0\', \'rated\\n  had an amazing mouth-watering ?chicken roll?worth every bite..a must try for every foodie?a variety of rolls and sandwiches are available as well .. a budding rafflesia ?\')]'

In [35]:
import re
rev2=re.sub('[^a-zA-Z]', ' ',rev)
rev2

'   rated        rated n  had an egg chicken roll and a paneer roll    really yummy    a must visit place    it would be good if you guys can keep the shop open on sunday s as well   good luck             rated        rated n  not just the roll but the filling tastes great  i would highly recommend to others to try their rolls  owners hospitality is also good and very friendly       rated        rated n  very nice place complete value for money   highly recommend must visit for any foodie   i would recommend the egg chicken roll    feeltheroll      rated        rated n  had an amazing mouth watering  chicken roll worth every bite  a must try for every foodie a variety of rolls and sandwiches are available as well    a budding rafflesia     '

In [36]:
rev3=re.sub('rated', ' ',rev2)
rev3

'              n  had an egg chicken roll and a paneer roll    really yummy    a must visit place    it would be good if you guys can keep the shop open on sunday s as well   good luck                        n  not just the roll but the filling tastes great  i would highly recommend to others to try their rolls  owners hospitality is also good and very friendly                  n  very nice place complete value for money   highly recommend must visit for any foodie   i would recommend the egg chicken roll    feeltheroll                 n  had an amazing mouth watering  chicken roll worth every bite  a must try for every foodie a variety of rolls and sandwiches are available as well    a budding rafflesia     '

In [37]:
rev4=re.sub('x',' ',rev3)
rev4

'              n  had an egg chicken roll and a paneer roll    really yummy    a must visit place    it would be good if you guys can keep the shop open on sunday s as well   good luck                        n  not just the roll but the filling tastes great  i would highly recommend to others to try their rolls  owners hospitality is also good and very friendly                  n  very nice place complete value for money   highly recommend must visit for any foodie   i would recommend the egg chicken roll    feeltheroll                 n  had an amazing mouth watering  chicken roll worth every bite  a must try for every foodie a variety of rolls and sandwiches are available as well    a budding rafflesia     '

    remove multiple spaces

In [38]:
rev5=re.sub(' +',' ',rev4)
rev5

' n had an egg chicken roll and a paneer roll really yummy a must visit place it would be good if you guys can keep the shop open on sunday s as well good luck n not just the roll but the filling tastes great i would highly recommend to others to try their rolls owners hospitality is also good and very friendly n very nice place complete value for money highly recommend must visit for any foodie i would recommend the egg chicken roll feeltheroll n had an amazing mouth watering chicken roll worth every bite a must try for every foodie a variety of rolls and sandwiches are available as well a budding rafflesia '

In [39]:
## takes 7-8 min
total_review=' '
for review in newdata['reviews_list']:
    review=review.lower()
    review=re.sub('[^a-zA-Z]', ' ',review)
    review=re.sub('rated', ' ',review)
    review=re.sub('x',' ',review)
    review=re.sub(' +',' ',review)
    total_review=total_review + str(review)
 

In [40]:
## generate wordcloud
import wordcloud
from wordcloud import WordCloud,STOPWORDS

In [41]:
stopwords=set(STOPWORDS)

In [ ]:
## takes 3 mins
wordcloud = WordCloud(width = 800, height = 800, 
                background_color ='white', 
                stopwords = stopwords, 
                min_font_size = 10).generate(total_review) 
# plot the WordCloud image                        
plt.figure(figsize = (8, 8)) 
plt.imshow(wordcloud) 
plt.axis("off") 

    Observations
    Of the 50 most frequent words across customer reviews, best are: place,food,taste,ambience,good..
    The only negative word in the top 50 is "bad".

In [ ]:
### but it is hard to estimate over here which words wins by how much??

    for this u can use plots/charts

In [359]:
from nltk.corpus import RegexpTokenizer as regextoken

In [360]:
# Converting all the text to lowercase
newdata["reviews_list"] = newdata["reviews_list"].apply(lambda x: x.lower())

'''
def apply_lower(x):
    return x.lower()
newdata["reviews_list"] = newdata["reviews_list"].apply(apply_lower)
   

'''

'\ndef apply_lower(x):\n    return x.lower()\n newdata["reviews_list"] = newdata["reviews_list"].apply(apply_lower)\n   \n\n'

In [361]:
##  Creating a regular expression tokenizer that have only alphabets , ie remove all the special characters
# This will return separate words (tokens) from the text in the form of list
tokenizer = regextoken("[a-zA-Z]+") 

In [362]:
tokenizer

RegexpTokenizer(pattern='[a-zA-Z]+', gaps=False, discard_empty=True, flags=<RegexFlag.UNICODE|DOTALL|MULTILINE: 56>)

In [36]:
newdata["reviews_list"][0]

'[(\'rated 5.0\', "rated\\n  had an egg chicken roll and a paneer roll... really yummy... a must visit place... it would be good if you guys can keep the shop open on sunday\'s as well.. good luck... :-)"), (\'rated 5.0\', \'rated\\n  not just the roll but the filling tastes great. i would highly recommend to others to try their rolls. owners hospitality is also good and very friendly.\'), (\'rated 4.5\', \'rated\\n  very nice place complete value for money ? highly recommend.must visit for any foodie . i would recommend the egg chicken roll . #feeltheroll\'), (\'rated 5.0\', \'rated\\n  had an amazing mouth-watering ?chicken roll?worth every bite..a must try for every foodie?a variety of rolls and sandwiches are available as well .. a budding rafflesia ?\')]'

In [363]:
## tokenize data of very first review
print(tokenizer.tokenize(newdata["reviews_list"][0]))

['rated', 'rated', 'n', 'had', 'an', 'egg', 'chicken', 'roll', 'and', 'a', 'paneer', 'roll', 'really', 'yummy', 'a', 'must', 'visit', 'place', 'it', 'would', 'be', 'good', 'if', 'you', 'guys', 'can', 'keep', 'the', 'shop', 'open', 'on', 'sunday', 's', 'as', 'well', 'good', 'luck', 'rated', 'rated', 'n', 'not', 'just', 'the', 'roll', 'but', 'the', 'filling', 'tastes', 'great', 'i', 'would', 'highly', 'recommend', 'to', 'others', 'to', 'try', 'their', 'rolls', 'owners', 'hospitality', 'is', 'also', 'good', 'and', 'very', 'friendly', 'rated', 'rated', 'n', 'very', 'nice', 'place', 'complete', 'value', 'for', 'money', 'highly', 'recommend', 'must', 'visit', 'for', 'any', 'foodie', 'i', 'would', 'recommend', 'the', 'egg', 'chicken', 'roll', 'feeltheroll', 'rated', 'rated', 'n', 'had', 'an', 'amazing', 'mouth', 'watering', 'chicken', 'roll', 'worth', 'every', 'bite', 'a', 'must', 'try', 'for', 'every', 'foodie', 'a', 'variety', 'of', 'rolls', 'and', 'sandwiches', 'are', 'available', 'as', 'w

In [38]:
## takes 2 mins
# Applying the tokenizer to each row of the reviews
review_tokens = newdata["reviews_list"].apply(tokenizer.tokenize)

In [39]:
# Examining the tokens created for the first row / restaurant
print(review_tokens[0])

['rated', 'rated', 'n', 'had', 'an', 'egg', 'chicken', 'roll', 'and', 'a', 'paneer', 'roll', 'really', 'yummy', 'a', 'must', 'visit', 'place', 'it', 'would', 'be', 'good', 'if', 'you', 'guys', 'can', 'keep', 'the', 'shop', 'open', 'on', 'sunday', 's', 'as', 'well', 'good', 'luck', 'rated', 'rated', 'n', 'not', 'just', 'the', 'roll', 'but', 'the', 'filling', 'tastes', 'great', 'i', 'would', 'highly', 'recommend', 'to', 'others', 'to', 'try', 'their', 'rolls', 'owners', 'hospitality', 'is', 'also', 'good', 'and', 'very', 'friendly', 'rated', 'rated', 'n', 'very', 'nice', 'place', 'complete', 'value', 'for', 'money', 'highly', 'recommend', 'must', 'visit', 'for', 'any', 'foodie', 'i', 'would', 'recommend', 'the', 'egg', 'chicken', 'roll', 'feeltheroll', 'rated', 'rated', 'n', 'had', 'an', 'amazing', 'mouth', 'watering', 'chicken', 'roll', 'worth', 'every', 'bite', 'a', 'must', 'try', 'for', 'every', 'foodie', 'a', 'variety', 'of', 'rolls', 'and', 'sandwiches', 'are', 'available', 'as', 'w

In [ ]:
### now from this above list,we will figure out we have some stopwords, it means we have to remove these stopwords like an,and,it etc

In [364]:
from nltk.corpus import stopwords

In [365]:
# These are common words defined by Python developers that typically don't add meaning to the text and can be removed
stop = stopwords.words("english")
print(stop)

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

In [ ]:
### we will figure out there are some more stopwords in my data, that we can add it to my stop list

In [366]:
# Adding custom words to stopwords 
stop.extend(["rated", "n", "nan", "x"])

In [237]:
print(stop)

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

In [238]:
### with respect to very first row, how to remove stopwords
rev=review_tokens[0]
print(rev)

['rated', 'rated', 'n', 'had', 'an', 'egg', 'chicken', 'roll', 'and', 'a', 'paneer', 'roll', 'really', 'yummy', 'a', 'must', 'visit', 'place', 'it', 'would', 'be', 'good', 'if', 'you', 'guys', 'can', 'keep', 'the', 'shop', 'open', 'on', 'sunday', 's', 'as', 'well', 'good', 'luck', 'rated', 'rated', 'n', 'not', 'just', 'the', 'roll', 'but', 'the', 'filling', 'tastes', 'great', 'i', 'would', 'highly', 'recommend', 'to', 'others', 'to', 'try', 'their', 'rolls', 'owners', 'hospitality', 'is', 'also', 'good', 'and', 'very', 'friendly', 'rated', 'rated', 'n', 'very', 'nice', 'place', 'complete', 'value', 'for', 'money', 'highly', 'recommend', 'must', 'visit', 'for', 'any', 'foodie', 'i', 'would', 'recommend', 'the', 'egg', 'chicken', 'roll', 'feeltheroll', 'rated', 'rated', 'n', 'had', 'an', 'amazing', 'mouth', 'watering', 'chicken', 'roll', 'worth', 'every', 'bite', 'a', 'must', 'try', 'for', 'every', 'foodie', 'a', 'variety', 'of', 'rolls', 'and', 'sandwiches', 'are', 'available', 'as', 'w

In [239]:
print([token for token in rev if token not in stop])

['egg', 'chicken', 'roll', 'paneer', 'roll', 'really', 'yummy', 'must', 'visit', 'place', 'would', 'good', 'guys', 'keep', 'shop', 'open', 'sunday', 'well', 'good', 'luck', 'roll', 'filling', 'tastes', 'great', 'would', 'highly', 'recommend', 'others', 'try', 'rolls', 'owners', 'hospitality', 'also', 'good', 'friendly', 'nice', 'place', 'complete', 'value', 'money', 'highly', 'recommend', 'must', 'visit', 'foodie', 'would', 'recommend', 'egg', 'chicken', 'roll', 'feeltheroll', 'amazing', 'mouth', 'watering', 'chicken', 'roll', 'worth', 'every', 'bite', 'must', 'try', 'every', 'foodie', 'variety', 'rolls', 'sandwiches', 'available', 'well', 'budding', 'rafflesia']


In [109]:
## takes 3 mins
## remove tokens from my entire data
review_tokens = review_tokens.apply(lambda x: [token for token in x if token not in stop])

In [ ]:
## using loops,own logic
''' review_tokens2=[]
for token in review_tokens:
    if token not in stop:
        review_tokens2.append(token)'''

In [110]:
type(review_tokens)

pandas.core.series.Series

In [111]:
# Concatenating all the reviews 
all_reviews = review_tokens.astype(str).str.cat()


# Concatenating all the reviews using own approach
## takes time
'''string=''
for i in range(len(review_tokens2)):
    string=string + ' '.join(review_tokens2[i])'''

"string=''\nfor i in range(len(review_tokens2)):\n    string=string + ' '.join(review_tokens2[i])"

In [112]:
all_reviews[0:200]

"['egg', 'chicken', 'roll', 'paneer', 'roll', 'really', 'yummy', 'must', 'visit', 'place', 'would', 'good', 'guys', 'keep', 'shop', 'open', 'sunday', 'well', 'good', 'luck', 'roll', 'filling', 'tastes'"

In [113]:
type(all_reviews)

str

In [ ]:
### now I have to convert this string into list as u will figure out , data inside is in the form of list actually bcz our main
### goal is to compute frequency of each word,so to obtain your goal, very first u have to convert your data in the form of list

In [114]:
len(all_reviews)

129341038

In [115]:
## takes 1 min
## perform tokenization to convert your string(all_reviews) into list
cleaned_reviews = tokenizer.tokenize(all_reviews)

In [116]:
len(cleaned_reviews)

13376224

In [117]:
type(cleaned_reviews)

list

In [118]:
print(cleaned_reviews[0:200])

['egg', 'chicken', 'roll', 'paneer', 'roll', 'really', 'yummy', 'must', 'visit', 'place', 'would', 'good', 'guys', 'keep', 'shop', 'open', 'sunday', 'well', 'good', 'luck', 'roll', 'filling', 'tastes', 'great', 'would', 'highly', 'recommend', 'others', 'try', 'rolls', 'owners', 'hospitality', 'also', 'good', 'friendly', 'nice', 'place', 'complete', 'value', 'money', 'highly', 'recommend', 'must', 'visit', 'foodie', 'would', 'recommend', 'egg', 'chicken', 'roll', 'feeltheroll', 'amazing', 'mouth', 'watering', 'chicken', 'roll', 'worth', 'every', 'bite', 'must', 'try', 'every', 'foodie', 'variety', 'rolls', 'sandwiches', 'available', 'well', 'budding', 'rafflesia', 'little', 'cafe', 'set', 'beautiful', 'location', 'ambiance', 'good', 'nthe', 'burger', 'filled', 'taste', 'chicken', 'nthe', 'banana', 'chocolate', 'sandwich', 'variety', 'would', 'never', 'find', 'restaurants', 'bangalore', 'drinks', 'wonderful', 'lot', 'flavours', 'choose', 'nit', 'took', 'little', 'time', 'expected', 'food

In [ ]:
# obtain the frequency of individual words in the reviews, for this u have to use FreqDist

In [89]:
from nltk import FreqDist, bigrams, trigrams

In [119]:
fd = FreqDist()

In [120]:
## takes 1 min
## checkout documentation by pressing Shift+Tab
for word in cleaned_reviews:
    fd[word]=fd[word]+ 1

In [121]:
# Examining the top 5 most frequent words
fd.most_common(5)

[('good', 328304),
 ('place', 317915),
 ('food', 307405),
 ('chicken', 141301),
 ('service', 109512)]

In [ ]:
## takes 1 min
# Plotting the top 50 most frequent words
plt.figure(figsize = (15, 8))
fd.plot(50)

In [ ]:
##Observations
## Of the 50 most frequent words across customer reviews, six reveal food preferences: chicken, biryani, veg, pizza, rice.

   Factors contributing to restaurant experience are mentioned in the following (descending) order of frequency:
   
   place > service > taste > ambience >time> staff > quality > menu  > quantity > friendly > delivery>.

### Bi-grams

In [ ]:
##Now let us repeat the analysis on a bi-gram level. Bi-grams are pairs of words that provide better insights than individual words.

In [126]:
from nltk import bigrams

In [127]:
# Generating bigrams from the reviews
bigrams = bigrams(cleaned_reviews)


In [128]:
bigrams

<generator object bigrams at 0x00000065B55470C8>

In [129]:
## takes 
# Getting the bigram frequency distribution
fd_bigrams = FreqDist()
for bigram in bigrams:
    fd_bigrams[bigram]=fd_bigrams[bigram] + 1
# Examining the top 5 most frequent bigrams
fd_bigrams.most_common(5)

[(('must', 'try'), 23182),
 (('really', 'good'), 22564),
 (('good', 'food'), 21589),
 (('food', 'good'), 21175),
 (('good', 'place'), 18690)]

In [ ]:
# Plotting the top 50 most frequent bigrams
plt.figure(figsize = (15, 8))
fd_bigrams.plot(50)

    Observations
    We have some new insights! Food items/preferences mentioned in the top 50 bigrams are non veg, ice cream, North Indian,
    chicken biryani
    Factors contributing to restaurant experience are : good food > good place > good service > value (for) money > pocket
    friendly > ambience good.

    There's a key insight here: the expense factor, which was missed by individual word frequency counts, was picked up by the bigram frequency counts.

In [131]:
fd_bigrams.most_common(50)

[(('must', 'try'), 23182),
 (('really', 'good'), 22564),
 (('good', 'food'), 21589),
 (('food', 'good'), 21175),
 (('good', 'place'), 18690),
 (('visit', 'place'), 15832),
 (('non', 'veg'), 13561),
 (('one', 'best'), 13537),
 (('main', 'course'), 13470),
 (('ice', 'cream'), 12961),
 (('nice', 'place'), 12348),
 (('also', 'good'), 11087),
 (('good', 'service'), 10902),
 (('must', 'visit'), 10327),
 (('service', 'good'), 9911),
 (('great', 'place'), 9419),
 (('north', 'indian'), 9419),
 (('visited', 'place'), 9232),
 (('taste', 'good'), 9170),
 (('place', 'good'), 8860),
 (('ambience', 'good'), 8789),
 (('value', 'money'), 8736),
 (('fried', 'rice'), 8427),
 (('ordered', 'chicken'), 8023),
 (('nwe', 'ordered'), 7978),
 (('quality', 'food'), 7860),
 (('chicken', 'biryani'), 7647),
 (('good', 'taste'), 7563),
 (('pocket', 'friendly'), 7513),
 (('pretty', 'good'), 7488),
 (('great', 'food'), 7351),
 (('zomato', 'gold'), 7340),
 (('good', 'ambience'), 7245),
 (('food', 'great'), 7163),
 (('f

    Zomato might also be happy to know their membership program "Zomato Gold" is in the top 50 bigrams, with 7340 mentions in the customer reviews.

### What about trigrams?

In [133]:
from nltk import trigrams

In [134]:
# Generating trigrams from the reviews
trigrams = trigrams(cleaned_reviews)

In [135]:
## takes 
fd_trigrams = FreqDist()
for trigram in trigrams:
    fd_trigrams[trigram] += 1

In [136]:

fd_trigrams.most_common(5)

[(('must', 'visit', 'place'), 3982),
 (('f', 'f', 'f'), 3916),
 (('veg', 'non', 'veg'), 3734),
 (('north', 'indian', 'food'), 3433),
 (('place', 'hangout', 'friends'), 2824)]

In [ ]:
plt.figure(figsize = (10, 5))
fd_trigrams.plot(50)


    Observations
    There appears to be some bad data (strings of "f f f") somewhere in the reviews, but we'll ignore that. The specific food preferences we can see here are north Indian Food, paneer butter masala, chicken fried rice, chicken ghee roast etc. Bangalore is full of chicken lovers...

On restaurant experience: a specific insight revealed by the trigrams is that many people are looking for places to hang out with their friends.

    We now have plenty of insights into customer preferences and experiences, and will move onto an analysis of Bangalore's cuisines.

In [240]:
# Creating a new dataset with cuisines info and restaurant ratings
cuisines = newdata[["cuisines", "rating"]]


In [241]:
cuisines.isnull().sum()

cuisines      23
rating      4524
dtype: int64

In [242]:
cuisines.dropna(inplace=True)

In [243]:
# Converting to lowercase
cuisines["cuisines"] = cuisines["cuisines"].astype(str).apply(lambda x: x.lower())

# Tokenizing the cuisines
cuisine_tokens = cuisines["cuisines"].apply(tokenizer.tokenize)


In [244]:
cuisine_tokens[0]

['fast', 'food']

In [245]:
type(cuisine_tokens)

pandas.core.series.Series

In [246]:
# Concatenating all the cuisine names into one text document
all_cuisines = cuisine_tokens.astype(str).str.cat()


In [247]:
type(all_cuisines)

str

In [248]:
all_cuisines[0:50]

"['fast', 'food']['fast', 'food', 'beverages']['fas"

In [249]:
### now I have to convert this string into list as u will figure out , data inside all_cuisines is in the form of list actually 
### & our main goal is to compute frequency of each word,so to obtain your goal, very first u have to convert all_cuisines in the form of list
cleaned_cuisines = tokenizer.tokenize(all_cuisines)



In [148]:
type(cleaned_cuisines)

list

In [149]:
# Generating cuisine frequencies 
fd_cuisine = FreqDist()
for cuisine in cleaned_cuisines:
    fd_cuisine[cuisine] += 1
    
# Printing the 50 most common cuisines (top 50)
print(fd_cuisine.most_common(5))

[('indian', 12398), ('north', 8662), ('chinese', 6283), ('food', 5263), ('south', 3706)]


In [ ]:
plt.figure(figsize=(15,9))
fd_cuisine.plot(50)

### Now start applying algo but before that u have to pre-process your data

In [153]:
newdata.columns

Index(['name', 'address', 'listed_in(type)_x', 'url', 'online_order',
       'book_table', 'rate', 'votes', 'phone', 'location', 'rest_type',
       'dish_liked', 'cuisines', 'approx_cost(for two people)', 'reviews_list',
       'menu_item', 'listed_in(type)_y', 'listed_in(city)', 'rating'],
      dtype='object')

In [367]:
# Converting all the text to strings

newdata[["reviews_list", "menu_item", "dish_liked", "cuisines"]] = newdata[["reviews_list", "menu_item", "dish_liked", "cuisines"]].astype("str")

# Combining all the text data into a single feature called "text"
newdata["text"] = newdata["reviews_list"] + " " + newdata["menu_item"] + " " + newdata["dish_liked"] + " " + newdata["cuisines"]

In [251]:
newdata.head()

name                                            address  \
0   #FeelTheROLL  Opposite Mantri Commercio, Outer Ring Road, De...   
1     #L-81 Cafe                          Sector 6, HSR Layout, HSR   
2     #L-81 Cafe                          Sector 6, HSR Layout, HSR   
3  #Vibes Restro  Marasur Gate, Chandapura - Anekal Road, Near A...   
4  #Vibes Restro  Marasur Gate, Chandapura - Anekal Road, Near A...   

                                   listed_in(type)_x  \
0                               [Delivery, Delivery]   
1  [Delivery, Dine-out, Delivery, Dine-out, Deliv...   
2  [Delivery, Dine-out, Delivery, Dine-out, Deliv...   
3                       [Buffet, Delivery, Dine-out]   
4                       [Buffet, Delivery, Dine-out]   

                                                 url online_order book_table  \
0  https://www.zomato.com/bangalore/feeltheroll-b...           No         No   
1  https://www.zomato.com/bangalore/l-81-cafe-hsr...          Yes         No   
2  https://www.zomato.com/bangalore/l-81-cafe-hsr...          Yes         No   
3  https://www.zomato.com/bangalore/vibes-restro-...           No         No   
4  https://www.zomato.com/bangalore/vibes-restro-...           No         No   

    rate  votes                             phone         location  \
0  3.4/5      7  +91 9108342079\r\n+91 9886117901        Bellandur   
1  3.9/5     48                    +91 9986210891              HSR   
2  3.9/5     48                    +91 9986210891              HSR   
3    NaN      0                    +91 9538911285  Electronic City   
4    NaN      0                    +91 9538911285  Electronic City   

       rest_type dish_liked                       cuisines  \
0    Quick Bites        nan                      Fast Food   
1    Quick Bites    Burgers           Fast Food, Beverages   
2    Quick Bites    Burgers           Fast Food, Beverages   
3  Casual Dining        nan  Continental, Chinese, Italian   
4  Casual Dining        nan  Continental, Chinese, Italian   

  approx_cost(for two people)  \
0                         200   
1                         400   
2                         400   
3                         700   
4                         700   

                                        reviews_list menu_item  \
0  [('rated 5.0', "rated\n  had an egg chicken ro...        []   
1  [('rated 4.0', 'rated\n  this little cafe is s...        []   
2  [('rated 4.0', 'rated\n  this little cafe is s...        []   
3  [('rated 5.0', "rated\n  great service and don...        []   
4  [('rated 5.0', "rated\n  great service and don...        []   

  listed_in(type)_y  listed_in(city) rating  \
0          Delivery        Bellandur      1   
1          Delivery              BTM      1   
2          Dine-out              BTM      1   
3            Buffet  Electronic City    NaN   
4          Delivery  Electronic City    NaN   

                                                text  
0  [('rated 5.0', "rated\n  had an egg chicken ro...  
1  [('rated 4.0', 'rated\n  this little cafe is s...  
2  [('rated 4.0', 'rated\n  this little cafe is s...  
3  [('rated 5.0', "rated\n  great service and don...  
4  [('rated 5.0', "rated\n  great service and don...

In [368]:
# Creating a new dataset with text and restaurant ratings
text_data = newdata[["text", "rating"]]


In [253]:
text_data.head()

text rating
0  [('rated 5.0', "rated\n  had an egg chicken ro...      1
1  [('rated 4.0', 'rated\n  this little cafe is s...      1
2  [('rated 4.0', 'rated\n  this little cafe is s...      1
3  [('rated 5.0', "rated\n  great service and don...    NaN
4  [('rated 5.0', "rated\n  great service and don...    NaN

    what all things, u have to apply on your text column, basically
    a.lowercase
    b.tokenization
    c.remove stopwords
    d.lemmatization

In [369]:
# Converting text to lowercase
text_data["text"] = text_data["text"].apply(lambda x: x.lower())


In [370]:
text_data.shape

(20915, 2)

In [371]:
df=text_data.sample(n=2000)

In [372]:
df.reset_index(inplace=True)

In [373]:
df.drop('index',axis=1,inplace=True)

In [374]:
df.head()

text rating
0  [('rated 5.0', 'rated\n  good music good ambie...      2
1  [('rated 3.0', 'rated\n  ordered paneer kabab ...    NaN
2  [('rated 5.0', "rated\n  amma's pastries is a ...      2
3           [] [] nan arabian, biryani, north indian    NaN
4  [('rated 1.0', 'rated\n  i packed a full meal ...      1

In [375]:
df.isnull().sum()

text        0
rating    418
dtype: int64

In [376]:
df.dropna(inplace=True)

In [377]:
df.shape

(1582, 2)

In [378]:
# Tokenizing the text as we have to remove stopwords from data
tokens = df["text"].apply(tokenizer.tokenize) 


In [379]:
type(tokens)

pandas.core.series.Series

In [ ]:
print(tokens[0])

In [381]:
len(tokens)

1582

In [382]:
## takes more than 2 min if considered entire data
# Removing stopwords 
tokens = tokens.apply(lambda x: [token for token in x if token not in stop])

In [383]:
from nltk.stem import WordNetLemmatizer

In [384]:
wl = WordNetLemmatizer()

In [ ]:
print(tokens[0])

In [386]:
data=tokens[0]

In [ ]:
print([wl.lemmatize(word) for word in data])

In [388]:
df['rating'].unique()

['2', '1', '0']
Categories (3, object): ['0' < '1' < '2']

In [389]:
df['rating'].dtype

CategoricalDtype(categories=['0', '1', '2'], ordered=True)

In [390]:
df['rating'].value_counts()

1    1113
0     270
2     199
Name: rating, dtype: int64

In [ ]:
# Applying the function to each row of the text
# i.e. reducing each word to its lemma

In [391]:
# Writing a function to lemmatize words

def lem(text):
    return [wl.lemmatize(word) for word in text]

In [392]:
## takes almost 10 second but if I considered entire data ,it will take time
tokens_new = tokens.apply(lem)

In [393]:
type(tokens_new)

pandas.core.series.Series

In [394]:
tokens_new.head()

0    [good, music, good, ambience, good, interior, ...
2    [amma, pastry, best, place, cake, pastry, choc...
4    [packed, full, meal, meal, received, rice, sam...
6    [ordered, following, cream, factory, michele, ...
9    [ordered, chicken, noodle, last, night, really...
Name: text, dtype: object

In [395]:
from sklearn.preprocessing import LabelEncoder

In [396]:
le=LabelEncoder()

In [397]:
df['rating']=le.fit_transform(df['rating'])

In [ ]:
print(tokens_new[0])

In [94]:
' '.join(tokens_new[0])

'one unique shawarma far nmore like kebab khubus light spice fresh nwhat make unique melt mouth kinda soft never something like ncost nlocation road parallel commercial street towards end nlandmark tiny little bakery opposite mosque bakery strong history decade around commercial street shopping post lunch looking place snack packed meat place thier mutton mince samosa mutton cutlet chicken bun mutton chicken puff must try trying get might tough task foot best option need munched standing something like thier piece mutton bone cutlet sometimes may prick throat except place good good food good embince good service good staff attitude really enjoyed good time friend college love kalmi kebab tandoori chicken happy say satisfied hospitality one old age eatery locality famous khara biscuit puff season serve haleem well sit place provided bon appetite chef bb nchefbb age old bakery situated meenaksi koil street ebrahim sahib street oph road junction probably among stood changing time hot sell

In [283]:
tokens_new.index

Int64Index([   0,    1,    2,    3,    4,    5,    6,    7,    8,    9,
            ...
            1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999],
           dtype='int64', length=1577)

In [399]:
## collect entire data into corpus as we have to NLP techniques to this corpus to convert it into some vectors
corpus=[]

for i in tokens_new.index:
    review=' '.join(tokens_new[i])
    corpus.append(review)
    

In [400]:
# Creating the TF_IDF model
from sklearn.feature_extraction.text import TfidfVectorizer
cv = TfidfVectorizer()
X = cv.fit_transform(corpus).toarray()

In [401]:
X.shape

(1582, 22643)

In [402]:
from sklearn.model_selection import train_test_split

In [403]:
# Splitting the data into train and test sets (stratified)
X_train, X_test, y_train, y_test = train_test_split(X, df['rating'], test_size = 0.3, random_state = 0,stratify=df['rating'])

In [404]:
from sklearn.linear_model import LogisticRegression

In [405]:
lr=LogisticRegression(multi_class='multinomial')

In [406]:
lr.fit(X_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='multinomial', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [407]:
# Make predictions.
predictions =lr.predict(X_test)

In [408]:
# Compute the error.
from sklearn.metrics import confusion_matrix
print(confusion_matrix(predictions, y_test))



[[  2   3   0]
 [ 79 326  41]
 [  0   5  19]]


In [409]:
from sklearn.metrics import accuracy_score
print(accuracy_score(predictions,y_test))
print('\n')

0.7305263157894737




In [410]:
#fit naive bayes
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier

In [411]:
### classifier models
models = []
models.append(('LogisticRegression', LogisticRegression()))
models.append(('Naive Bayes',GaussianNB()))
models.append(('RandomForest', RandomForestClassifier()))
models.append(('Decision Tree', DecisionTreeClassifier()))
models.append(('KNN', KNeighborsClassifier(n_neighbors = 5)))

In [412]:
# Make predictions on validation dataset

for name, model in models:
    print(name)
    model.fit(X_train, y_train)
    
    # Make predictions.
    predictions = model.predict(X_test)

    # Compute the error.
    from sklearn.metrics import confusion_matrix
    print(confusion_matrix(predictions, y_test))

    from sklearn.metrics import accuracy_score
    print(accuracy_score(predictions,y_test))
    print('\n')

LogisticRegression
[[  2   3   0]
 [ 79 326  41]
 [  0   5  19]]
0.7305263157894737


Naive Bayes
[[ 16  37   3]
 [ 48 215  28]
 [ 17  82  29]]
0.5473684210526316


RandomForest
[[  4   3   0]
 [ 77 327  39]
 [  0   4  21]]
0.7410526315789474


Decision Tree
[[ 24  62   1]
 [ 52 251  24]
 [  5  21  35]]
0.6526315789473685


KNN
[[ 10  37   3]
 [ 69 256  24]
 [  2  41  33]]
0.6294736842105263




In [ ]:
### cross validating my Logistic Regression model

In [413]:
from sklearn.model_selection import cross_val_score

In [414]:
## takes 2 mins
lr = LogisticRegression()
print(cross_val_score(lr, X, df['rating'], cv=5, scoring ='accuracy').mean())

0.7452621491035419
